test/functional/lib/tests.cpp

```diff
@@ -4358,6 +4817,344 @@ GenericRxFromTxInspect(
+
+    //
+    // Forward all packets.
+    //
+    XDP_RULE Rule;
+    Rule.Match = XDP_MATCH_ALL;
+    Rule.Action = XDP_PROGRAM_ACTION_L2FWD;
+    auto ProgramHandle =
+        CreateXdpProg(If.GetIfIndex(), &XdpInspectRxL2, If.GetQueueId(), XDP_GENERIC, &Rule, 1);
+
+    UINT Seed = (UINT)std::time(NULL);
+    TraceVerbose("Seeding %u", Seed);
+    std::srand(Seed);
+
+    for (UINT32 i = 0; i < 10000; i++) {
+        GENERIC_RX_FRAGMENT_PARAMS Params = {0};
+        std::vector<UINT32> SplitIndexes;
+        UINT8 SplitCount;
+        UINT8 TcpOptions[TH_MAX_LEN - sizeof(TCP_HDR)];
+
+        //
+        // Periodically advertise new NIC capabilities.
+        //
+        if ((i % 100) == 0) {
+            NDIS_OFFLOAD_PARAMETERS OffloadParams;
+            FN_OFFLOAD_OPTIONS OffloadOptions;
+
+            InitializeOffloadParameters(&OffloadParams);
+            FnMpInitializeOffloadOptions(&OffloadOptions);
+
+            if (std::rand() & 1) {
+                OffloadParams.LsoV2IPv4 = NDIS_OFFLOAD_PARAMETERS_LSOV2_ENABLED;
+                OffloadParams.LsoV2IPv6 = NDIS_OFFLOAD_PARAMETERS_LSOV2_ENABLED;
+            }
+
+            if (std::rand() & 1) {
+                OffloadParams.IPv4Checksum = NDIS_OFFLOAD_PARAMETERS_TX_RX_ENABLED;
+                OffloadParams.UDPIPv4Checksum = NDIS_OFFLOAD_PARAMETERS_TX_RX_ENABLED;
+                OffloadParams.UDPIPv6Checksum = NDIS_OFFLOAD_PARAMETERS_TX_RX_ENABLED;
+                OffloadParams.TCPIPv4Checksum = NDIS_OFFLOAD_PARAMETERS_TX_RX_ENABLED;
+                OffloadParams.TCPIPv6Checksum = NDIS_OFFLOAD_PARAMETERS_TX_RX_ENABLED;
+            }
+
+            OffloadOptions.GsoMaxOffloadSize = (std::rand() % MaxPayload) + 1;
+
+            //
+            // Set the offload and replace the offload resetter without
+            // triggering an immediate reset for the old handle (if any) here.
+            //
+            *OffloadReset.addressof() =
+                MpUpdateTaskOffload(
+                    GenericMp, FnOffloadCurrentConfig, &OffloadParams, &OffloadOptions).release();
+        }
+
+        Params.Action = XDP_PROGRAM_ACTION_L2FWD;
+        Params.TcpOptionsLength = (std::rand() % (sizeof(TcpOptions) + 1) / 4) * 4;
+        Params.PayloadLength =
+            (UINT16)(std::rand() % (MaxPayload - sizeof(TCP_HDR) - Params.TcpOptionsLength + 1));
+        Params.Backfill = std::rand() % 100;
+        Params.Trailer = std::rand() % 100;
+        Params.GroSegCount = (std::rand() % (Params.PayloadLength + TCP_HEADER_BACKFILL(Af))) + 1;
+        Params.LowResources = std::rand() & 1;
+
+        if ((std::rand() % 4 == 0)) {
+            Params.DataTrailer = std::rand() % (MaxDataTrailer + 1);
+        }
+
+        std::vector<UCHAR> Payload((SIZE_T)Params.PayloadLength + Params.DataTrailer);
+        std::generate(Payload.begin(), Payload.end(), []{ return (UCHAR)std::rand(); });
+
+        std::vector<UCHAR> PacketBuffer(
```

Comment:
Is using a vector better than `unique_malloc_ptr`? There's some subtle differences between std::vector and karray, where I had to replace `std::vector<UCHAR>` by `unique_malloc_ptr` in #489. In karray, IIRC, random access is only available after adding elements to the array.

Reply:
I was hoping to merge this before the STL removal 😆

I don't have a strong opinion on which class to use; I'm sure this code will eventually be replaced with karray and we'll be able to find a way to make it do what we need. If using a `unique_malloc_ptr` now would avoid karray troubles in the future, I can replace it now. Some other vector code in these tests does use the `append()` function, though, which is not directly compatible with a malloc_ptr.

Reply:
Up to you. Not a big deal. It would definitely save some of my time converting this to malloc in that STL removal PR :)

Reply:
Sure, I'll convert what I can - I'll leave the code relying on append semantics to be converted to karray.

Reply:
Actually, these all look like they should become karray - it would be awkward to use the unique_malloc_ptr to refer to an array of things.